src/Symfony/Component/HttpClient/UriTemplateHttpClient.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Component\HttpClient\UriTemplate\CallbackUriTemplateExpander;
+use Symfony\Component\HttpClient\UriTemplate\UriTemplateExpanderInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+class UriTemplateHttpClient implements HttpClientInterface
+{
+    private HttpClientInterface $client;
+    private UriTemplateExpanderInterface $expander;
+
+    private array $variables = [];
+
+    public function __construct(
+        HttpClientInterface $client = null,
+        UriTemplateExpanderInterface $expander = null,
+    ) {
+        $this->client = $client ?? HttpClient::create();
+        $this->expander = $expander ?? CallbackUriTemplateExpander::createFromPopularVendors();
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if ($this->variables && null !== ($options['base_uri'] ?? null)) {
+            $options['base_uri'] = $this->expander->expand($options['base_uri'], $this->variables);
+        }
+
+        return $this->client->request(
+            $method,
+            $this->variables ? $this->expander->expand($url, $this->variables) : $url,
+            $options,
+        );
+    }
+
+    public function stream(iterable|ResponseInterface $responses, float $timeout = null): ResponseStreamInterface
+    {
+        return $this->client->stream($responses, $timeout);
+    }
+
+    public function withOptions(array $options): static
+    {
+        if ($this->variables && null !== ($options['base_uri'] ?? null)) {
+            $options['base_uri'] = $this->expander->expand($options['base_uri'], $this->variables);
```

Comment:
Do we really want to expand the base URI? I'm not sure it makes sense to me.
Instead, what about adding an expandUri() method to the client?
I'd also suggest making this class use DecoratorTrait

Reply:
>  Instead, what about adding an expandUri() method to the client?

And it would need to be used as the value of the `base_uri` option?
```php
$this->client->withVariables([/*stuff*/])->request('GET', '/foo', [
    'base_uri' => $this->client->withVariables([/*stuff*/])->expandUri('https://{foo}/{bar}',
]);
```
That seems redundant.

Reply:
no need for any redundancy:
`$this->client->request('GET', 'https://{foo}/{bar}/foo', ['vars' => ...]);`